server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java

```diff
@@ -648,6 +650,13 @@ public IndexSearcher searcher() {
         return searcher;
     }
 
+    public FieldInfos getFieldInfos() {
+        if (this.fieldInfos == null) {
+            this.fieldInfos = searcher == null ? FieldInfos.EMPTY : FieldInfos.getMergedFieldInfos(searcher.getIndexReader());
```

Comment:
if searcher is null, this means we're not on a shard, and calling this method should be illegal?